File: kotlin/src/sort/QuickSort.kt
package sort

import kotlin.random.Random

class QuickSort {
    fun sort(list: List<Int>): List<Int> {
        if (list.size < 2) return list

        val x = pickPivot(list)

        val listSmallerThenX = sort(list.filter { y -> y < x })
        val listGreaterThenX = sort(list.filter { y -> y > x })

        return listSmallerThenX + listOf(x) + listGreaterThenX
    }

    private fun pickPivot(list: List<Int>): Int {
        return list[Random.nextInt(list.size)]
    }
}

File: dart/lib/sort/quick_sort.dart
import 'dart:math';

class QuickSort {
  List<int> sort(List<int> list) {
    if (list.length < 2) return list;

    final x = _pickPivot(list);

    final listSmallerThanX = sort(list.where((y) => y < x).toList());
    final listGreaterThanX = sort(list.where((y) => y > x).toList());

    return listSmallerThanX + [x] + listGreaterThanX;
  }

  int _pickPivot(List<int> list) => list[Random().nextInt(list.length)];
}
